daraja/src/main/java/com/github/daraja/utils/Utils.kt

```diff
@@ -39,3 +46,44 @@ fun getPassword(businessShortCode: String, passkey: String, timestamp: String):
     // encode the password to Base64
     return Base64.encodeToString(str.toByteArray(), Base64.NO_WRAP)
 }
+
+suspend fun <T> safeApiCall(
+    dispatcher: CoroutineDispatcher,
+    apiCall: suspend () -> T
+): Resource<T> {
+    return withContext(dispatcher) {
+        try {
+            Resource.Success(apiCall.invoke())
+        } catch (throwable: Throwable) {
+            Timber.e(throwable)
+            when (throwable) {
+                is IOException -> Resource.Error()
+                is HttpException -> {
+                    val stringErrorBody = errorBodyAsString(throwable)
+                    if (stringErrorBody != null) {
+                        val errorResponse = convertStringErrorResponseToJsonObject(stringErrorBody)
+                        Resource.Error(
+                            errorMessage = errorResponse?.errorMessage,
+                            throwable = throwable
+                        )
+                    } else {
+                        Resource.Error(null, null)
+                    }
+                }
+                else -> {
+                    Resource.Error(null, null)
+                }
+            }
+        }
+    }
+}
+
+private fun convertStringErrorResponseToJsonObject(jsonString: String): ErrorResponse? {
+    val gson = Gson()
+    return gson.fromJson(jsonString, ErrorResponse::class.java)
+}
+
+fun errorBodyAsString(throwable: HttpException): String? {
```

Comment:
Let's make this function private since its only used in this file.

Reply:
Sure will do that